core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/ByteBuffersAsyncRequestBody.java

```diff
@@ -0,0 +1,136 @@
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.internal.util.Mimetype;
+import software.amazon.awssdk.utils.BinaryUtils;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * An implementation of {@link AsyncRequestBody} for providing data from the supplied {@link ByteBuffer} array. This is created
+ * using static methods on {@link AsyncRequestBody}
+ *
+ * @see AsyncRequestBody#fromBytes(byte[])
+ * @see AsyncRequestBody#fromByteBuffer(ByteBuffer)
+ * @see AsyncRequestBody#fromByteBuffers(ByteBuffer...)
+ * @see AsyncRequestBody#fromString(String)
+ */
+@SdkInternalApi
+public final class ByteBuffersAsyncRequestBody implements AsyncRequestBody {
+    private static final Logger log = Logger.loggerFor(ByteBuffersAsyncRequestBody.class);
+
+    private final String mimetype;
+    private final Long length;
+    private final ByteBuffer[] buffers;
+
+    private ByteBuffersAsyncRequestBody(String mimetype, Long length, ByteBuffer... buffers) {
+        this.mimetype = mimetype;
+        this.length = length;
+        this.buffers = buffers;
+    }
+
+    @Override
+    public Optional<Long> contentLength() {
+        return Optional.ofNullable(length);
+    }
+
+    @Override
+    public String contentType() {
+        return mimetype;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super ByteBuffer> s) {
+        // As per rule 1.9 we must throw NullPointerException if the subscriber parameter is null
+        if (s == null) {
+            throw new NullPointerException("Subscription MUST NOT be null.");
+        }
+
+        // As per 2.13, this method must return normally (i.e. not throw).
+        try {
+            s.onSubscribe(
+                new Subscription() {
+                    private final AtomicInteger index = new AtomicInteger(0);
+                    private final AtomicBoolean completed = new AtomicBoolean(false);
+
+                    @Override
+                    public void request(long n) {
+                        if (completed.get()) {
+                            return;
+                        }
+
+                        if (n > 0) {
+                            int i = index.getAndIncrement();
+
+                            if (i >= buffers.length) {
+                                return;
+                            }
+
+                            long remaining = n;
+
+                            do {
+                                ByteBuffer buffer = buffers[i];
+                                if (!buffer.hasArray()) {
+                                    buffer = ByteBuffer.wrap(BinaryUtils.copyBytesFrom(buffer));
+                                }
```

Comment:
Is this because we have assumptions downstream of this publisher that the buffer `hasArray`?

Reply:
I've done this to maintain how the existing code will convert direct buffers (off heap) to non-direct (on heap) buffers see: https://github.com/aws/aws-sdk-java-v2/issues/3928
I've tested with the Netty and CRT clients and neither seem to have any issues handling off-heap buffers but I was attempting to separate out the zero-copy functionality to future PR's

Reply:
> I've done this to maintain how the existing code will convert direct buffers (off heap) to non-direct (on heap) buffers see: https://github.com/aws/aws-sdk-java-v2/issues/3928
> I've tested with the Netty and CRT clients and neither seem to have any issues handling off-heap buffers but I was attempting to separate out the zero-copy functionality to future PR's

Got it, I agree it makes sense to keep it separate for now.